File: QuangNinhA_2324/Bai1/frac.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iomanip>
#include <limits>
#include <fstream>  // Thêm thư viện này

using namespace std;

// Hàm tính tổng phân số
double calculateFraction(double x, double y, double z, double w) {
    return (x / y) + (z / w);
}

// Hàm tìm thứ tự tối ưu
vector<int> minFractionOrder(int a, int b, int c, int d) {
    vector<int> numbers = {a, b, c, d};
    vector<int> bestOrder;
    double minValue = numeric_limits<double>::infinity();

    // Duyệt qua tất cả các hoán vị
    do {
        int x = numbers[0];
        int y = numbers[1];
        int z = numbers[2];
        int w = numbers[3];

        // Tính tổng phân số
        double totalValue = calculateFraction(x, y, z, w);

        if (totalValue < minValue) {
            minValue = totalValue;
            bestOrder = {x, y, z, w};
        }
    } while (next_permutation(numbers.begin(), numbers.end()));

    return bestOrder;
}

int main() {
    // Đọc dữ liệu từ tệp
    ifstream input("frac.inp");
    ofstream output("frac.out");

    int a, b, c, d;
    input >> a >> b >> c >> d;

    // Tìm thứ tự tối ưu
    vector<int> result = minFractionOrder(a, b, c, d);

    // Ghi kết quả ra tệp
    for (int i = 0; i < result.size(); i++) {
        output << result[i] << (i < result.size() - 1 ? " " : "");
    }
    output << endl;

    return 0;
}

File: THCS/2023_2024/QuangNinhA_2324/Bai1/frac.py
from itertools import permutations

def min_fraction_order(a, b, c, d):
    numbers = [a, b, c, d]
    best_order = None
    min_value = float('inf')

    # Duyệt qua tất cả các hoán vị
    for perm in permutations(numbers):
        x, y, z, w = perm
        
        # Tính tổng phân số
        total_value = (x / y) + (z / w)
        
        if total_value < min_value:
            min_value = total_value
            best_order = perm

    return best_order

if __name__ == "__main__":
    # Đọc dữ liệu từ tệp
    with open("frac.inp", "r") as file:
        a, b, c, d = map(int, file.readline().strip().split())

    # Tìm thứ tự tối ưu
    result = min_fraction_order(a, b, c, d)

    # Ghi kết quả ra tệp
    with open("frac.out", "w") as file:
        file.write(" ".join(map(str, result)) + "\n")
